### PR TITLE
Convert pythonfinder from pydantic to vanilla dataclasses

### PR DESCRIPTION
**Pull request by matteius**
Ok -- This is it folks, the pythonfinder PR you've all been waiting for: removal of pydantic.  I've ported the equivalent changes over to pipenv and got the tests passing there (and ported those changes back to here).  

Corresponding pipenv PR that shows these changes passing:  https://github.com/pypa/pipenv/actions/runs/7676466722/job/20924059607

Key highlights:
* Remove the posix path conversions and learn more heavily into pathlib.
* Use string path keys to the paths dictionary
* reduce complexity in the python version comparison operators (and fix at least one bug there).
* Probably some other things--you should want to review this  PR.

### CODE REVIEW

**Comment by matteius** [comment]
Initial pass is actually working for mac os and linux but needs some investigation on the windows side:  https://github.com/pypa/pipenv/actions/runs/7606560585/job/20712546096

File: src/pythonfinder/models/mixins.py [hunk 1]
```diff
@@ -25,51 +22,51 @@
...
+    is_executable_ref: bool | None = None
+    is_python_ref: bool | None = None
+
+    def __post_init__(self):
+        if not self.children_ref:
+            self._gen_children()
+
+    def set_children(self, children):
+        # If children are not provided, generate them
+        return children or self._gen_children()
```
Comment (oz123) [inline src/pythonfinder/models/mixins.py:51]: Seems like a potential bug is hiding here.
`children` is a container, `self._gen_children()` populates `PathEntry.children_ref`, but it does not return any value.

**Review by oz123 — CHANGES_REQUESTED** [review]

File: src/pythonfinder/models/python.py [hunk 2]
```diff
@@ -405,23 +390,25 @@ def version_tuple(self) -> tuple[int, int, int, bool, bool, bool]:
...
-        python_name: str | None = None,
-    ) -> bool:
+        major=None,
+        minor=None,
+        patch=None,
+        pre=False,
+        dev=False,
+        arch=None,
+        debug=False,
+        python_name=None,
```
[the remark this answers is not in the document]
Comment (matteius) [reply src/pythonfinder/models/python.py:400]: The downsides of using LLMs on an air-plane I guess, I'll restore them.

File: src/pythonfinder/models/mixins.py (hunk 1 above)
Comment (matteius) [reply src/pythonfinder/models/mixins.py:51]: Huh, nothing is actually using this -- going to remove.

**Review by oz123 — APPROVED** [review]

**Pull request merged**